core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionRepository.java

```diff
@@ -47,15 +56,70 @@ public FunctionImplementation compile(FunctionName functionName, List<Expression
    * Resolve the {@link FunctionBuilder} in Builtin Function Repository.
    *
    * @param functionSignature {@link FunctionSignature}
-   * @return {@link FunctionBuilder}
+   * @return Original function builder if it's a cast function or all arguments have expected types.
+   *         Otherwise wrap its arguments by cast function as needed.
    */
   public FunctionBuilder resolve(FunctionSignature functionSignature) {
     FunctionName functionName = functionSignature.getFunctionName();
     if (functionResolverMap.containsKey(functionName)) {
-      return functionResolverMap.get(functionName).resolve(functionSignature);
+      Pair<FunctionSignature, FunctionBuilder> resolvedSignature =
```

Comment:
Feels the cast logic, data type and function signature feels not expressed in the abstract manner. No clear thoughts now.

Reply:
Got your point. Checking if we can hide this logic in function signature and data type abstraction underlying. Thanks!

Reply:
Checked but found the problem lies in compiling CAST function. Once the casting logic is moved out of current `BuiltinFunctionRepository`, the `compile` method is inaccessible. Need to figure out better approach and may need to do major refactoring on higher level (ex. adding additional symbol layer to avoid compiling so early). Thanks!